community/io/src/test/java/org/neo4j/adversaries/pagecache/AdversarialReadPageCursor.java

```diff
@@ -255,24 +367,39 @@ public void close()
     @Override
     public boolean shouldRetry() throws IOException
     {
-        adversary.injectFailure( FileNotFoundException.class, IOException.class, SecurityException.class,
+        state.injectFailure( FileNotFoundException.class, IOException.class, SecurityException.class,
                 IllegalStateException.class );
-        if ( currentReadIsInconsistent )
+        if ( state.hasPreparedInconsistentRead() )
+        {
+            resetDelegate();
+            return true;
+        }
+        if ( state.hasInconsistentRead() )
         {
-            currentReadIsInconsistent = false;
-            delegate.shouldRetry();
-            delegate.setOffset( 0 );
+            resetDelegate();
             return true;
         }
         boolean retry = delegate.shouldRetry();
         return retry || (linkedCursor != null && linkedCursor.shouldRetry());
     }
 
+    private void resetDelegate() throws IOException
+    {
+        delegate.shouldRetry();
+        delegate.setOffset( 0 );
+        delegate.checkAndClearBoundsFlag();
+        delegate.clearCursorError();
+    }
+
     @Override
     public int copyTo( int sourceOffset, PageCursor targetCursor, int targetOffset, int lengthInBytes )
     {
-        adversary.injectFailure( IndexOutOfBoundsException.class );
-        if ( !currentReadIsInconsistent )
+        state.injectFailure( IndexOutOfBoundsException.class );
+        if ( state.currentReadIsPreparingInconsistent )
```

Comment:
this if looks like state abstraction gives a leak and require external manipulation over internal state